src/Interpreters/InterpreterCheckQuery.cpp

```diff
@@ -17,63 +18,158 @@ namespace DB
-        auto path_column = block_structure[0].type->createColumn();
-        auto is_passed_column = block_structure[1].type->createColumn();
-        auto message_column = block_structure[2].type->createColumn();
 
-        for (const auto & check_result : check_results)
+        if (is_check_completed)
+            return {};
+
+        auto status = worker_result.wait_for(std::chrono::milliseconds(100));
+        is_check_completed = (status == std::future_status::ready);
+
+        if (is_check_completed)
         {
-            path_column->insert(check_result.fs_path);
-            is_passed_column->insert(static_cast<UInt8>(check_result.success));
-            message_column->insert(check_result.failure_message);
+            worker_result.get();
+            auto result_block = getBlockFromCheckResult(check_results, check_query_single_value_result);
+            check_results.clear();
+            return Chunk(result_block.getColumns(), result_block.rows());
         }
 
-        block = Block({
-            {std::move(path_column), block_structure[0].type, block_structure[0].name},
-            {std::move(is_passed_column), block_structure[1].type, block_structure[1].name},
-            {std::move(message_column), block_structure[2].type, block_structure[2].name}});
+        std::lock_guard lock(mutex);
+        progress(progress_rows, 0);
+        progress_rows = 0;
+
+        if (check_query_single_value_result || check_results.empty())
+        {
+            return Chunk();
+        }
+
+        auto result_block = getBlockFromCheckResult(check_results, check_query_single_value_result);
+        check_results.clear();
+        return Chunk(result_block.getColumns(), result_block.rows());
     }
 
-    BlockIO res;
-    res.pipeline = QueryPipeline(std::make_shared<SourceFromSingleChunk>(std::move(block)));
+private:
+    void worker()
+    {
+        table->checkData(query_ptr, context,
+            [this](const CheckResult & check_result, size_t new_total_rows)
+            {
+                if (isCancelled())
+                    return false;
+
+                std::lock_guard lock(mutex);
+                if (new_total_rows > total_rows)
+                {
+                    addTotalRowsApprox(new_total_rows - total_rows);
+                    total_rows = new_total_rows;
+                }
+                progress_rows++;
+
+                if (!check_result.success)
+                {
+                    LOG_WARNING(&Poco::Logger::get("InterpreterCheckQuery"),
+                        "Check query for table {} failed, path {}, reason: {}",
+                        table->getStorageID().getNameForLogs(),
+                        check_result.fs_path,
+                        check_result.failure_message);
+                }
+
+                check_results.push_back(check_result);
+
+                bool should_continue = check_result.success || !check_query_single_value_result;
```

Comment:
I assumed it was good to have a mode with early exit and thought we could do it since the user couldn't know how many errors there were with single-value result. However, its doubtful, so I returned it to the original behavior.